### DISCUSSION TITLE
Programmatically interacting with git-lfs

### DISCUSSION THREAD

**Discussion by alexflint** [git-lfs/git-lfs#5732] [General]
I am trying to write a Go program that interacts via REST APIs with git-lfs. I have Go code that creates git trees (and later commits and branches) via the github API like this:

```go
newBlob, _, err := a.github.Git.CreateBlob(ctx, orgName, repoName, &github.Blob{
    Content:  github.String(base64EncodedFileContents),
    Encoding: github.String("base64"),
})

newTree, _, err := a.github.Git.CreateTree(ctx, orgName, repoName, mainBranchSHA, []*github.TreeEntry{{
    Mode: github.String("100644"),
    Path: github.String("path/in/repository"),
    SHA:  newBlob.SHA,
}})
```

I would like to push certain of these blobs to git-lfs and store the appropriate pointer information in git, and I would like to do this via REST APIs, avoiding checking out the whole repository locally, and avoiding calling out to separate executables.

I see in the top-level readme of this repository that you warn:

> Please do not import the git-lfs module into other Go code and do not rely on it as a source code dependency.

I also have found the Batch API docs here: https://github.com/git-lfs/git-lfs/blob/main/docs/api/batch.md

Are there any libraries that would help me interact directly with the Batch API from a Go program? For experimental work, are there any parts of this repository that I might make use of for that purpose?

My goal is to create blobs in git-lfs and reference them as appropriate in git blobs/trees/commits, using Go, without checking out the whole repository into the filesystem and without calling out to any separate executable.

Any pointers whatsoever would be much appreciated!

**Comment by chrisd8088** [comment]
Hey, thanks for the question!

As you've already noted, we don't provide a stable Go language API within the `git-lfs` project, although for legacy reasons it's registered as a "module" in the Go ecosystem.  Rather, we expect users to use the compiled binaries for the Git LFS client.

That said, it's an open-source project, so you're of course able to pick and choose any bits of the source code you want (so long as you provide appropriate attribution in any published results according to our license, obviously).  In particular, you might want to look at a few files like `tq/api.go`, `tq/transfer.go`, and the JSON schemas in `tq/schemas/`.

Good luck with your project, and let us know how it goes!

**Reply by alexflint** [reply]
I got it working! I'm able to upload and download individual files from git-lfs, using ssh to discover the URLs and to get the auth headers. I'll post the code to a github repo and send a link here, as it might be helpful to have as a bare-bones / single-file client implementation.
